[build-script] XCTest CI builds on OS X

modocache opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?
<!-- Description about pull request. -->


Fix the OS X XCTest CI preset:

1. Installation of XCTest isn't supported on OS X. After all, the library is meant to be used on Linux, and not intended to be used or installed on OS X. Therefore, no longer pass an `INSTALL_PATH` to the `xcodebuild` invocation on OS X.
2. `xcodebuild` expects `swift-stdlib-tool` to be present in the same directory as the `SWIFT_EXEC` it's given. Re-use some shellscript from the toolchain installation script to copy the stdlib-tool substitute.

This allows the XCTest CI to work on all platforms, and should unblock https://github.com/apple/swift-corelibs-xctest/pull/77. That pull request will make it much easier for new contributors to start working on swift-corelibs-xctest.

#### Resolved bug number: None
<!-- If this pull request resolves any bugs from Swift bug tracker -->

* * * *

<!-- This selection should only be completed by Swift admin -->
Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please smoke test
        OS X platform               | @swift-ci Please smoke test OS X platform
        Linux platform              | @swift-ci Please smoke test Linux platform

 **Validation Testing**

        Platform     | Comment
        ------------ | -------------
        All supported platforms     | @swift-ci Please test
        OS X platform               | @swift-ci Please test OS X platform
        Linux platform              | @swift-ci Please test Linux platform

Note: Only members of the Apple organization can trigger swift-ci.
</details>
<!-- Thank you for your contribution to Swift! -->

tkremenek commented:
@swift-ci test

modocache commented:
Looks like some Swift package tests are failing due to something related to SwiftPM. Doesn't seem related to this PR. @kostiakoval @aciidb0mb3r are you aware of any ongoing CI issues in SwiftPM?

gottesmm commented:
Also, do we have ci tests in GitHub for swiftpm? How did these breakages escape at least this sort of testing?

Sent from my iPhone

> On Apr 2, 2016, at 2:07 PM, Brian Gesiak <notifications@github.com> wrote:
> 
> Looks like some Swift package tests are failing due to something related to SwiftPM. Doesn't seem related to this PR. @kostiakoval @aciidb0mb3r are you aware of any ongoing CI issues in SwiftPM?
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
>

modocache commented:
@gottesmm I believe the problem is that package tests are only run for apple/swift CI. The preset used by SwiftPM CI doesn't include them. That may be reasonable--they are quite expensive.

/cc @shahmishal for CI wisdom.

gribozavr commented on utils/build-script-impl:
```diff
@@ -1929,20 +1938,26 @@ for deployment_target in "${HOST_TARGET}" "${CROSS_TOOLS_DEPLOYMENT_TARGETS[@]}"
-                        SWIFT_LINK_OBJC_RUNTIME=YES \
-                        SKIP_INSTALL=NO \
-                        DEPLOYMENT_LOCATION=YES \
-                        DSTROOT="${XCTEST_BUILD_DIR}" \
```
If you drop `DSTROOT`, where would `xcodebuild` place the build products?

modocache replied:
The idea here is that the Swift build script no longer minds where the build products are placed. I'd like for the build script to be able to build and test XCTest to make it easier for newcomers to contribute (everyone uses the same invocation to test). But attempting to have it install build products in a specific location was a misstep on my part: we don't support installing corelibs-xctest on OS X, so we shouldn't try to place build products in a specific directory either.

With this change, build products are placed in the usual Xcode location: the derived data folder.

Alternatively, it might be possible to respect `DSTROOT` here, but I just can't get the build to work when doing so... :sweat_smile:

gribozavr replied:
One good reason to continue to pass `DSTROOT` is to allow us to reliably clean the build directory on CI.

modocache replied:
I could invoke `xcodebuild clean build` here, if you'd prefer.

gottesmm commented:
@modocache We are trying to move towards a model where all commits have some sort of CI testing. SwiftPM is no exception to this rule.

If the tests are too expensive then the tests should be sped up or a reasonable subset should be provided that can be run.

modocache commented:
@gottesmm I can definitely sympathize with your argument, especially because I've accidentally broken the package tests before in #1704. Again, I'll leave it up to @shahmishal to determine what's the best approach here.

@shahmishal Any thoughts on @gribozavr's concerns about derived data above? I'm excited to get this merged, it'll be a big improvement in XCTest's CI.

Also, if someone could try the tests again, that'd be much appreciated! :bow:

shahmishal commented:
@modocache I agree with @gribozavr we should specify derived data dir, so we can reliably clean the workspace. 

What is the error are you seeing with `DSTROOT="${XCTEST_BUILD_DIR}"`?

shahmishal commented:
@swift-ci Please test

modocache commented:
> What is the error are you seeing with `DSTROOT="${XCTEST_BUILD_DIR}"`?

Oops, sorry, the error I was seeing must have been due to `INSTALL_PATH` and the other parameters I removed. I added `DSTROOT` and everything works fine. Thanks!!

I believe this is now good to merge? The CI passed last time it was run.

modocache commented:
@shahmishal Friendly ping! I think this is a big improvement for corelibs-xctest development.

gribozavr commented:
@swift-ci Please test and merge

modocache commented:
Just my luck: two CI failures and both appear unrelated:

- [OS X #782](https://ci.swift.org/job/swift-PR-osx/782/console): The `XCTAssert` failure fixed in https://github.com/apple/swift/commit/2d50e813478f5962b6942a7eb581f4d70e2ab3eb.
- [Linux #1003](https://ci.swift.org/job/swift-PR-Linux/1003/console): `ClangModules/attr-swift_name_renaming.swift` test failure, seems unrelated to this.

modocache commented:
Excellent!! Thanks!
